src/goto-programs/assign_params_as_non_det.cpp

```diff
@@ -36,24 +57,217 @@ bool assign_params_as_non_det::runOnFunction(
       return false; // Not expected
     exprt lhs = symbol_expr(*context.find_symbol(_id));
 
-    // rhs
-    exprt rhs = exprt("sideeffect", lhs.type());
-    rhs.statement("nondet");
-    rhs.location() = l;
+    typet l_t = lhs.type();
+    if (l_t.is_pointer() && l_t.subtype() != empty_typet())
+    {
+      /*
+        e.g. int* lhs;
+        to
+        lhs = null;
+        bool temp;
+        temp = nondet();
+        if(temp)
+        {
+          int temp2;
+          temp2 = nondet();
+          assume(temp2 != 0);
+          lhs = &temp2;
+        }
+        
+        During this process we create two auxiliary variable. One for the boolean flag, and one for the object.
+        - We need to consider the situation where pointers can be null, thus we put the assignment under an if-statement
+        - If the flag (`temp`) is true, we create an object (`temp2`) and assign its address to the pointer.
+      */
+
+      // lhs = null;
+      exprt zero_rhs = gen_zero(l_t);
+      zero_rhs.location() = l;
+
+      // assignment
+      goto_programt zero_program;
+      goto_programt::targett zero_assignment =
+        zero_program.add_instruction(ASSIGN);
+      zero_assignment->location = l;
+      zero_assignment->function = it->location.get_function();
+
+      code_assignt zero_assign(lhs, zero_rhs);
+      zero_assign.location() = it->location;
+      expr2tc new_zero_assign;
+      migrate_expr(zero_assign, new_zero_assign);
+      zero_assignment->code = new_zero_assign;
+
+      // insert
+      goto_program.insert_swap(it++, *zero_assignment);
+      --it;
+
+      // get subType() => int
+      typet subt = l_t.subtype();
+      // if it's symbol, get the original type
+      if (subt.is_symbol())
+        subt = context.find_symbol(subt.identifier())->type;
+
+      // create obj and move it to the symbol table
```

Comment:
I am not sure if this is your case, but what if the parameter is an array given as a pointer? 
```c
void sort(int *ptr, int n) { 
  ptr[1] = ptr[2];
}
```

I guess this could be sorted out in the future using an infinite array but for now could you please add some KNOWNBUG test cases (if they actually are)?

Reply:
I did some tests and it seems it works fine. 

So if there is a pointer parameter, we do
```c++
void sort(int *ptr) { 
  // e.g. int x; sort(&x);
  // or int x[10]; sort(x);
  ptr = nondet();  // this should cover the array case?
  bool ptr#temp_bool;
  ptr#temp_bool = nondet();
  if(!ptr#temp_bool)
  {
     int ptr#temp;  //This would make more sense for struct type as we are creating an object
     ptr#temp = nondet();
     assume(ptr#temp);
     ptr = &ptr#temp;
  }
}

```